File: dclab-template-tirocinio/03-code/matlab/Codice1.m
% Esempio: calcolo iterativo del gradiente proiettato
function x = projected_gradient(grad, project, x0, step, tol, max_iter)
if nargin < 4, step = 0.01; end
if nargin < 5, tol = 1e-6; end
if nargin < 6, max_iter = 1000; end

x = x0;
for k = 1:max_iter
    x_new = project(x - step * grad(x));
    if norm(x_new - x) < tol
        x = x_new;
        return
    end
    x = x_new;
end
end

File: dclab-template-beamer/03-code/python/Codice1.py
# Esempio: risoluzione numerica di un problema di ottimizzazione
# vincolata con il metodo del gradiente proiettato.

import numpy as np


def projected_gradient(grad, project, x0, step=0.01, tol=1e-6, max_iter=1000):
    """Minimizza una funzione tramite gradiente proiettato.

    grad: funzione che restituisce il gradiente nel punto x
    project: funzione di proiezione sul dominio ammissibile
    x0: punto iniziale
    """
    x = x0
    for _ in range(max_iter):
        x_new = project(x - step * grad(x))
        if np.linalg.norm(x_new - x) < tol:
            return x_new
        x = x_new
    return x


if __name__ == "__main__":
    grad = lambda x: 2 * x - np.array([4.0, 4.0])
    project = lambda x: np.clip(x, 0, None)  # vincolo x >= 0

    x_star = projected_gradient(grad, project, x0=np.zeros(2))
    print(f"Soluzione trovata: {x_star}")
